rust/connlib/clients/android/src/lib.rs

```diff
@@ -381,6 +409,16 @@ fn connect(
     };
     let session = Session::connect(args, runtime.handle().clone());
 
+    runtime.spawn({
+        let session = session.clone();
+
+        async move {
+            while let Some(tun) = new_tun_rx.recv().await {
+                session.set_tun(tun);
+            }
+        }
```

Comment:
Maybe we can make it so that we discard packets after calling `on_set_interface` but before it's set again.

Reply:
Hmm, maybe interior-mutability would be a better approach?

Reply:
> Now the eventloop thread might write to the tun interface, I think this might be a problem since that might be a closed file descriptor (I remember that it might be an exception?)

Writing to a closed TUN device should just give us an IO error right? So we might see a few warnings in the logs but it should fix itself as soon as the new device is set?

I am debugging some other issues with Android right now though so maybe that is related.

Reply:
When writing to a closed FD the process recieves a SIGPIPE signal which (I might be misremembering this) causes the program to crash (see EPIPE here: https://linux.die.net/man/2/write)

I think--, reading a bit on it now, we can handle explicitly SIGPIPE to prevent it from crashing.

Before when went single-threaded we used to have an `AtomicBool` that told us if the FD was closed or open without having to lock a `Mutex` for each operation, though I don't recommend to do this again, a Mutex is better until a benchmark says otherwise.

Reply:
> Now the eventloop thread might write to the tun interface, I think this might be a problem since that might be a closed file descriptor (I remember that it might be an exception?)
> 
> So we might need a sync mechanism on `Tun` to prevent this.

We need some more refactoring but I think we can eventually solve this once we completely get rid of the callbacks:

- Instead of a callback, connlib will be emitting an event that the interface needs to be updated.
- The Android eventloop can synchronously execute the FFI callback, obtain the new FD and set the new interface on the tunnel.

Reply:
Yeah, that sounds like the way to go, so for the moment we need a `Mutex` or something